### DISCUSSION TITLE
How are you supposed to install a script from file?

### DISCUSSION THREAD

**Discussion by rthfhgyhy567** [violentmonkey/violentmonkey#2092] [Q&A]
I don't see any option in VM to install a script from file.  All of these scripts that are in there were imported from Tamper Monkey, which does have that option.  There is no way I can even install a user script locally from my device.  I can't even paste the code into a new script as shown in this video because it glitches out.  Please add an option to import a script from file to make this process work much better.

https://github.com/violentmonkey/violentmonkey/assets/104237805/1f9f44f8-a049-4a9d-9a43-72dff76ac2a6

**Comment by tophf** [comment]
It's a bug, obviously, and you're the first one to report it, so either no one ever used this feature in a phone or it worked for everyone. Pasting works for me in an Android emulator.

What is your version of Violentmonkey? Browser and version? Android version?

@gera2ld, could you test it on a live device?

**Reply by gera2ld** [reply]
Yeah I can reproduce it. My guess is that the keyboard on mobile behaves differently from that on desktop and CodeMirror cannot deal with it, even "backspace" doesn't work properly.

**Reply by gera2ld** [reply]
It seems to be a known issue that CodeMirror 5 [doesn't](https://github.com/codemirror/codemirror5/issues/5678) support mobile devices well. [CodeMirror 6](https://codemirror.net/) has made an effort to improve support for mobile devices. I guess we need to switch to 6 if we want to support mobile devices.

**Reply by tophf** [reply]
I'd prefer we backport the fix, because CM6 is overly complicated and its improvements are overshadowed for me by its janky scrolling. If we change the editor I'd prefer a light version of Monaco.

**Reply by gera2ld** [reply]
> I'd prefer we backport the fix

I don't think it's possible since it's a complete rewrite.

> I'd prefer a light version of Monaco

I agree with that, but I could not find a light version :joy:

**Reply by tophf** [reply]
I meant just the pasting itself. It should be something isolated.

**Reply by gera2ld** [reply]
Or we can just implement "Install from file" which I think is better than hacking pasting.

**Reply by tophf** [reply]
Yeah. Will you do it or me? BTW 8c7be64f breaks the dropdown menu, it becomes clipped inside the header's area due to overflow-y:hidden.

**Reply by gera2ld** [reply]
Thanks for the feedback, let me fix it. I will add the "Install from file" menu as well.

**Reply by tophf** [reply]
I remembered that the file `input` element's `accept` should use MIME types e.g. `'application/javascript, text/plain'` instead of file extensions in order to support file choosers in Android e.g. GDrive.

**Reply by tophf** [reply]
Hmm, I think [CM6 fixes pasting](https://github.com/codemirror/view/blob/5d4d490fea204e7a152121062d5d5d256d136063/src/input.ts#L701) by changing `Text` to `text/plain` and optionally using a temporary textarea.

**Comment by rthfhgyhy567** [comment]
I'm not sure what the violent monkey version was because this made me rage uninstall, but it was downloaded a few months ago.  What exactly is the bug you are referring to?  The lack of the import from file option or the clip board not working?

    Kiwi Browser V124.0.6327.4
    Android 13

**Reply by tophf** [reply]
The bug is about pasting. There's no `install from file` button because it's not necessary in a desktop version and we didn't think it might be convenient in a mobile version. I guess we can add it both in a desktop and mobile version into the `+` menu in the dashboard.

**Reply by rthfhgyhy567** [reply]
I think it is much easier to install a script from file instead of having to open the script with a text editor and copy and paste the code.  Plus, there is probably a character limit for clipboards on android, so there is a chance that parts of the code get cut off with that method.

**Reply by tophf** [reply]
On a mobile platform, yes. On a desktop you can drag'n'drop the file into the Violentmonkey dashboard, which is why there's no such button, but we can still add it on both platforms for consistency.

**Reply by kunkanader** [reply]
drag'n'drop is not as intuitive as a dedicated button for many users and more importantly it doesn't allow to edit scripts that have small minor errors (like wrong dependency link) or install them without editing out these errors as well. Thus forcing user to edit such script in a dedicated text editor before drag'n'drop'ing it again

**Reply by tophf** [reply]
You can edit them, see https://violentmonkey.github.io/posts/how-to-edit-scripts-with-your-favorite-editor/

**Comment by Crayolable** [comment]
Loading the page in desktop mode will allow you to paste your code.

**Comment by shodanx2** [comment]
Hello, 
First time violentmonkey user,
This page is the first google link for the question 
"violent monkey how to add file on disc"

So, I am trying to add this file I made called "claude_usage_tracker.user.js"

<img width="2850" height="1614" alt="image" src="https://github.com/user-attachments/assets/7aca4dda-e948-4570-a2d5-2d2598d064cc" />

There is an import from zip but that doesn't see the file.

I would rather avoid copying file contents and use the file itself, that is the reason why the file abstraction exists, I am hoping there will be a place where I can enable disable such files from a checkboxed list somewhere ? Not just smoosh everything into one custom.js file ? 

I when you press "import from zip", you get this

<img width="505" height="442" alt="image" src="https://github.com/user-attachments/assets/3aca1755-0e8b-4b71-b1cc-dafb537e22a4" />

It kind of feels like, whatever you import here, is going to overwrite every other script too ? So .... is there any way to have say 3 script file, and enable and disable them individually with a single click and no copy pasting around of stuff ?

**Reply by shodanx2** [reply]
So, if you do paste the files, then there is an interface to enable/disable scripts individually.
It's just that you can feed files or packages of files ?

<img width="1483" height="936" alt="image" src="https://github.com/user-attachments/assets/0eef2032-bb4c-47d0-9a32-275dacd0e426" />

Unless you open each of them, and copy and paste the content into the editor one at a time.

I think this could be made more user friendly ...

Also, how do I share my script once it works ?
Is there a share button somewhere ?

I can't seem to find a way to put it on the violentmonkey store. I don't want to put it on the tampermonkey store if there is one.

**Reply by tophf** [reply]
Simply drag'n'drop the file from the explorer into any Violentmonkey page.
